nh_clinical/activity_extension.py

```diff
@@ -538,110 +530,241 @@ def get_activity_user_ids(self, cr, uid, activity_id, context=None):
+        if triggered_context and location_id:
+            location = location_model.browse(location_id)
+            # Create a list of bools representing if the activity and location
+            # contexts match. If none of them match then return False
+            if not any(
+                    [c.name == triggered_context for c in
+                     location.context_ids]):
+                return False
+        return True
+
+    @api.model
+    def check_trigger_domains(self, activity, domains=None):
+        """
+        Check if any records are returned for the supplied domains. If so then
+        return False
+
+        :param activity: nh.clinical.activity instance
+        :param domains: list of dicts with object (model to query) and domain
+            (search query to carry out on model)
+        :return: True if no records found, False is records found
+        :rtype: bool
+        """
+        if not domains:
+            return False
+        for domain in domains:
+            domain_pool = self.env[domain['object']]
+            search_domain = domain['domain'] + [
+                ['parent_id', '=', activity.spell_activity_id.id]
+            ]
+            if domain_pool.search(search_domain):
+                return True
+        return False
+
+    @api.model
+    def trigger_policy_cancel_others(self, spell_activity_id, trigger_model):
+        """
+        Cancel all open activities when triggering the policy
+
+        :param spell_activity_id: ID of the spell's activity record. This is
+            used by the cancel_
+        """
+        activity_model = self.env['nh.activity']
+        model_data = self.env['ir.model.data']
+        # Create a dict of the model names and cancel reasons so can do a
+        # lookup while also leaving room to extend with other models
+        reasons = {
+            'nh.clinical.patient.placement': model_data.get_object(
+                'nh_clinical', 'cancel_reason_placement').id
+        }
+        activity_model.cancel_open_activities(
+            spell_activity_id,
+            trigger_model,
+            cancel_reason_id=reasons.get(self._name)
+        )
+
+    def _get_policy_create_data(self, case=None):
+        """
+        Get the create_data dictionary for the policy
+
+        :param case: Case that controls the logic used when getting policy
+        :type case: int
+        :return: Dictionary of values to use when creating a new activity
+        :rtype: dict
+        """
+        return {}
+
+    @staticmethod
+    def trigger_policy_create_activity(source_activity, model, case=None):
+        """
+        Create a new activity using the data defined in the policy with the
+        source activity as it's creator
+
+        :param source_activity: Activity to use to get activity info from
+        :type source_activity: nh.activity record
+        :param model: Model to create activity on
+        :type model: Odoo model
```

Comment:
Could be more specific by putting `openerp.models.Model`

Reply:
changed